Implemented proven transaction validation

polydez opened the pull request:
Implemented `validate` method for `ProvenTransaction`. Automatically validating `ProvenTransaction` during building and deserializing.

Resolves https://github.com/0xPolygonMiden/miden-base/issues/528

hackaugusto reviewed (approved):
LGTM

bobbinth commented on miden-tx/src/error.rs:
```diff
@@ -95,6 +95,66 @@ impl Display for TransactionProverError {
+// ================================================================================================
+
+#[derive(Debug)]
+pub enum ProvenTransactionError {
+    AccountFinalHashMismatch(Digest, Digest),
+    AccountIdMismatch(AccountId, AccountId),
+    InputNotesError(TransactionInputError),
+    NoteDetailsForUnknownNotes(Vec<NoteId>),
+    OffChainAccountWithDetails(AccountId),
+    OnChainAccountMissingDetails(AccountId),
+    NewOnChainAccountRequiresFullDetails(AccountId),
+    ExistingOnChainAccountRequiresDeltaDetails(AccountId),
+    OutputNotesError(TransactionOutputError),
+    ProofVerificationError(TransactionVerifierError),
+}
```
nit: let's alphabetize error variants.

bobbinth commented on miden-tx/src/proven_tx/mod.rs:
```diff
@@ -123,6 +113,75 @@ impl ProvenTransaction {
+            }
+        }
+
+        if !skip_proof_checking {
+            let verifier = TransactionVerifier::new(MIN_PROOF_SECURITY_LEVEL);
+            verifier.verify(self).map_err(ProvenTransactionError::ProofVerificationError)?;
+        }
```
So - if we want to verify the proof, we need to move `ProvenTransaction` to `miden-tx` crate - right? Are there any other reasons to move besides that?

polydez replied:
Yes, I had to move `ProvenTransaction` to `miden-tx` because of references to `miden-tx` and `miden-lib`.

bobbinth reviewed (commented):
Thank you! Looks good! Not a full review yet - but I left a couple of comments/questions inline.

polydez commented:
After discussion with @bobbinth we decided to remove proof checking during tx validation. I reverted the most of changes here and just added `validate` method to `ProvenTransaction` and use it during building and deserialization of `ProvenTransaction`.

bobbinth commented on objects/src/transaction/proven_tx.rs:
```diff
@@ -124,6 +125,70 @@ impl ProvenTransaction {
         self.block_ref
     }
+
+    fn validate(&self) -> Result<(), ProvenTransactionError> {
```
Let's add a section separator above this line (similar to what we have on lines 71 - 72) with title "HELPER METHODS".

Also, I wonder if we should change the signature to:
```
fn validate(self) -> Result<Self, ProvenTransactionError>
```
This way, we'd be able to do `proven_transaction.validate()` on line 334 below and won't need to add extra `Ok(proven_transaction)`.

But this is a very minor nit and it is fine to leave as is as well.

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple of tiny nits inline - but it is good to go even as is.
